### ISSUE TITLE
FAN53555

### ISSUE THREAD

**Issue by xypron** [VenKamikaze/firefly-rk3399-build#1]
@VenKamikaze 

To fix the issue with the FAN53555 driver the following patch is sufficient in the 4.16.13 kernel:
https://lkml.org/lkml/2018/6/4/781

Best regards

Heinrich

**Comment by VenKamikaze** [comment]
Thanks, however this doesn't solve the issue I have with the pwm_regulator module on my board. Was your patch enough to solve all issues with your board on boot?

**Comment by xypron** [comment]
@VenKamikaze 

https://lkml.org/lkml/2018/6/4/781 is sufficient to boot.

The following error messages are shown by dmesg:
```
rockchip_mmc_get_phase: invalid clk rate
rockchip_mmc_get_phase: invalid clk rate
rockchip_mmc_get_phase: invalid clk rate
rockchip_mmc_get_phase: invalid clk rate
rk_iommu ff8f3f00.iommu: Failed to get clk 'iface': -2
rk_iommu ff903f00.iommu: Failed to get clk 'iface': -2
```

For iommu I have seen a patch in linux next https://patchwork.kernel.org/patch/10250747/ .

You can find how I build the v4.17 kernel at https://github.com/xypron/kernel-firefly-rk3399.

**Comment by Dvergatal** [comment]
@VenKamikaze

Have you succeded with pwm_regulator?

**Comment by xypron** [comment]
I have not touched the board since 2018. My kernel patch was rejected.

**Comment by Dvergatal** [comment]
@xypron
that was question to VenKamikaze but as i understand you also have firefly-rk3399?

**Comment by VenKamikaze** [comment]
@Dvergatal 

I have a working kernel and device tree spec for my FireFly 2GB RK3399 board, but without WiFi or Bluetooth. I found that the WiFi+BT additions made to the DTS on mainline a while back caused my board to hang on boot. I investigated a little but I was too far out of my depth to resolve the issues and ended up just pulling those bits out of the DTS to get a bootable board again.

I also suspect there to be variants of the FireFly RK3399 board, so I can't guarantee what I've done will work for everyone. For instance, xypron's patches did not work for my board.

The last kernel + DTS I compiled for my board was for kernel 5.3. You can find a DTS and some compiled images that worked for me here: https://github.com/VenKamikaze/firefly-rk3399-build/tree/master/kernel/configs/working-5.3-mainline/

I haven't checked recently, but it should be possible to get newer kernels going as well - the changes needed for the 4.16 DTS (when I first got something going) to the 5.3 DTS were quite similar, other than having to remove the extra WiFi+BT changes.

**Comment by Dvergatal** [comment]
@VenKemikaze
Yeah it is also working for me:] and thx to your kernel configuration i have discovered a bug in u-boot mainline dts files for firefly. U-Boot was missing regulator-init-microvolt in rk3399-firefly-u-boot.dtsi   an kernel was freezing during boot process. By the way there is no wifi APXXXX driver in 5.3 kernel. It is only in rockchip 4.4 line.

**Comment by VenKamikaze** [comment]
That's great! I'm happy that this helped someone.

Did you happen to get a custom u-boot going with the board? 

Thanks for letting me know about the wifi too.

**Comment by Dvergatal** [comment]
Yes, i used mainline u-boot tpl/spl but it had a bug which i have fixed with help of kevery and this is my [patch](http://u-boot.10912.n7.nabble.com/PATCH-Missing-regulator-init-microvolt-in-rk3399-firefly-u-boot-dtsi-td434199.html). I have also newest arm trusted firmware working with it with op-tee 3.7.0.

**Comment by Dvergatal** [comment]
I'm sorry the wifi driver name is AP6XXX. My mistake.
